lsp/src/service/types.ts

```diff
@@ -0,0 +1,18 @@
+import {
+    CompletionItem,
+    CompletionList,
+    Diagnostic,
+    Hover,
+    HoverParams,
+    TextDocumentPositionParams
+} from 'vscode-languageserver'
+import { TextDocument } from 'vscode-languageserver-textdocument'
+
+export interface LanguageService {
+    completion: (
+        document: TextDocument,
+        textDocumentPosition: TextDocumentPositionParams
+    ) => Promise<CompletionItem[] | CompletionList>
+    diagnostic: (document: TextDocument) => Promise<Diagnostic[]>
+    hover: (document: TextDocument, params: HoverParams) => Promise<Hover | null>
```

Comment:
We should add comments to this interface and methods to signal its importance to contributors, and so that they understand the contract they are implementing. Consider adding links to the LSP spec for reference too if there is a (close to) 1:1 mapping with these methods.